### PR TITLE
core: introduce "blockdev@.target" that "systemd-cryptsetup@.service" may use to properly order themselves for mount and swap units

### PR DESCRIPTION
**Pull request by poettering**
lacks documentation still

### CODE REVIEW

**Comment by fbuihuu** [comment]
I still don't understand what you meant by "people have stacked stuff though, such a global ordering is just sloppy and problematic in those cases." but anyways..

I did give a testing for you and it's still broken:

```
journalctl -b -1 -p warning
Dec 20 09:37:13 localhost systemd-cryptsetup[1577]: Device cr_test is still in use.
Dec 20 09:37:13 localhost systemd-cryptsetup[1577]: Failed to deactivate: Device or resource busy
Dec 20 09:37:13 localhost systemd[1]: systemd-cryptsetup@cr_test.service: Failed with result 'exit-code'.
```

As already pointed out somewhere else, the device name used in crypttab can different from the one used in fstab.

**Comment by poettering** [comment]
> I still don't understand what you meant by "people have stacked stuff though, such a global ordering is just sloppy and problematic in those cases." but anyways..

let's say you have a crypto device, and on top of that a file system, and on top of that another crypto device and on top of that another DM device and another fs, and another crypto device. With your patch you'd order all crypto devices early and all file systems late. Which obviously can't model the setup, because in this case it's not that easy that all crypto devices come first and all mounts after that. You'd get dep loops galore...

With my patch we have blockdev@.target, which is specifically insrted between just the cryptsetup for a device and the user of the device, hence a stack like the above actually works.

And yes, people use such complex stacks. The setup described in the very recent #14386 is a general purpose distro image with a loopback device in the middle of its storage stack...

Yeah, my approach doesn't work if you work heavily with aliases and reference your devices in crypttab and in fstab by unrelated names, because then we can't relate them to each other. But I think a scheme that requires you to use consisting naming is still way better than using a dependency logic that cannot express many setups *at* *all*.

I mean, the DM volume names are explicitly chosen by the user, I think it's not too strange to expect people to use the same name when using these virtual block devices. Sure, it's something to document, but it's not totally strange.

What's the precise testcase you had issues with? What#s the crypttab line? what the fstab line?

**Comment by poettering** [comment]
Also, please pass the "reviewed/needs-rework" label when there's something actionable, i.e. a code review actually done or so.

**Comment by fbuihuu** [comment]
> let's say you have a crypto device, and on top of that a file system, and on top of that another crypto device and on top of that another DM device and another fs, and another crypto device. With your patch you'd order all crypto devices early and all file systems late. Which obviously can't model the setup, because in this case it's not that easy that all crypto devices come first and all mounts after that. You'd get dep loops galore...

OK I see the use case now.

But it's currently not working at least when crypttab and fstab are used, is it ?

**Comment by fbuihuu** [comment]
> Yeah, my approach doesn't work if you work heavily with aliases and reference your devices in crypttab and in fstab by unrelated names, because then we can't relate them to each other. But I think a scheme that requires you to use consisting naming is still way better than using a dependency logic that cannot express many setups _at_ _all_.

Well using using UUID in crypttab and fstab is common practice and wide spread.

> What's the precise testcase you had issues with? What#s the crypttab line? what the fstab line?

Nothing fancy really:

```
# cat /etc/crypttab
cr_test UUID=195a967d-a373-4b53-a53e-91c82169b345
```
```
# cat /etc/fstab
UUID=53c6f477-3392-4f7c-a2a6-df23092eee4c /mnt ext4 defaults 0 0
```

As you can see it's simply based on UUIDs.

**Comment by fbuihuu** [comment]
> Yeah, my approach doesn't work if you work heavily with aliases and reference your devices in crypttab and in fstab by unrelated names, because then we can't relate them to each other. But I think a scheme that requires you to use consisting naming is still way better than using a dependency logic that cannot express many setups _at_ _all_.

Well using using UUID in crypttab and fstab is common practice and wide spread.

> What's the precise testcase you had issues with? What#s the crypttab line? what the fstab line?

Nothing fancy really:

```
# cat /etc/crypttab
cr_test UUID=195a967d-a373-4b53-a53e-91c82169b345
```
```
# cat /etc/fstab
UUID=53c6f477-3392-4f7c-a2a6-df23092eee4c /mnt ext4 defaults 0 0
```

As you can see it's simply based on UUIDs.

**Comment by poettering** [comment]
so the UUID use in crypttab is fine. the one in fstab is problematic though because we cannot determine the reltationship between the line and the one in crypttab until the moment the device pops up. THus we cannot know that the crypttab line makes it available and thus we also don't determine that we need to detach the crypt device only after unmounting the mount point...

The only way we could somehow salvage that if we'd dynamically update the deps, i.e. add the moment the cryptsetup device was activated update the deps towards the blockdev target.

File: man/systemd.special.xml [hunk 1]
```diff
@@ -845,6 +846,23 @@
...
 
       <variablelist>
+        <varlistentry>
+          <term><filename>blockdev@.target</filename></term>
+          <listitem><para>This template unit may be used to order mount units and other consumers of block
+          devices against services that synthesize these block devices. This is intended to be used to order
+          storage services (such as
+          <citerefentry><refentrytitle>systemd-cryptsetup@.service</refentrytitle><manvolnum>5</manvolnum></citerefentry>)
+          that allocate and manage a virtual block device against mount units and other consumers of
+          it. Specifically, the storage services are supposed to be orderd before an instance of
```
Comment (keszybz) [inline man/systemd.special.xml:856]: There's some typos here, but we can fix that later.

**Pull request merged**